File: scm-csharp/scheme/primitives/PrimitiveHashTableSetComparatorB.cs
namespace scheme;

public class PrimitiveHashTableSetComparatorB : Primitive
{
    public override string Name() => "%hash-table-set-comparator!";
    public override string Info() => "";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 2, 2);
        if (arguments[0] is not SchemeHashTable ht)
            throw new SchemeError(pos, "%hash-table-set-comparator!: not a hash table");
        ht.Comparator = arguments[1];
        return Value.NIL;
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveHashTableSetComparatorB.java
package scheme.primitives;

import scheme.*;

public class PrimitiveHashTableSetComparatorB extends Primitive {
    @Override
    public String name() { return "%hash-table-set-comparator!"; }

    @Override
    public String info() { return ""; }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 2, 2);
        if (!(arguments[0] instanceof SchemeHashTable))
            throw new SchemeError(pos, name() + ": not a hash table");
        SchemeHashTable ht = (SchemeHashTable) arguments[0];
        ht.comparator = arguments[1];
        return Value.NIL;
    }
}
